src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -817,209 +817,380 @@ private BoundCollectionExpression ConvertCollectionExpression(
+
+                            var collectionBuilderMethods = _binder.GetCollectionBuilderMethods(syntax, namedType, _diagnostics, forParams: false);
+                            if (collectionBuilderMethods.IsEmpty)
+                                return null;
+
+                            (collectionCreation, collectionBuilderMethod, collectionBuilderElementsPlaceholder) = bindCollectionBuilderInfo(
+                                _binder, _node, _targetType, collectionBuilderMethods, _diagnostics);
+                            if (collectionCreation is null || collectionBuilderMethod is null || collectionBuilderElementsPlaceholder is null)
+                                return null;
+
+                            var lastParameterType = collectionBuilderMethod.Parameters.Last().Type;
+                            Debug.Assert(lastParameterType.IsReadOnlySpan());
+                            elementType = ((NamedTypeSymbol)lastParameterType).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
                         }
+                        break;
 
-                        var lastParameterType = collectionBuilderMethod.Parameters.Last().Type;
-                        Debug.Assert(lastParameterType.IsReadOnlySpan());
-                        elementType = ((NamedTypeSymbol)lastParameterType).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
-                    }
-                    break;
+                }
 
-                case CollectionExpressionTypeKind.ImplementsIEnumerable:
-                    if (targetType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Immutable_ImmutableArray_T), TypeCompareKind.ConsiderEverything))
+                var elements = _node.Elements;
+                var builder = ArrayBuilder<BoundNode>.GetInstance(elements.Length);
+                BoundObjectOrCollectionValuePlaceholder? implicitReceiver = null;
+
+                if (collectionTypeKind is CollectionExpressionTypeKind.ImplementsIEnumerable)
+                {
+                }
+                else
+                {
+                    if (_node.WithElement != null)
                     {
-                        diagnostics.Add(ErrorCode.ERR_CollectionExpressionImmutableArray, syntax, targetType.OriginalDefinition);
-                        return BindCollectionExpressionForErrorRecovery(node, targetType, inConversion: true, diagnostics);
+                        if (collectionTypeKind is CollectionExpressionTypeKind.ArrayInterface)
+                        {
+                            Debug.Assert(constructor is null, "Only initialized in ImplementsIEnumerable case in Conversions.GetCollectionExpressionConversion.");
+                            collectionCreation = BindCollectionArrayInterfaceConstruction(
+                                _targetType, list_T__ctor, list_T__ctorInt32, _node.WithElement);
+                        }
```

Comment:
It looks like this code is no longer reachable, but it looks like it used to do something meaningful.